File: swift/Sources/DuetShells/Relay.swift
// Copyright (c) 2026 Modaal.dev
// Licensed under the MIT License. See LICENSE file for details.

// Extracted from the reference adopter’s navigation shells: breaks the
// child-environment ↔ parent-store construction cycle without IUO closures. The
// composition root creates the relay, hands `send` to the child environment it is
// building, and assigns `sink` once the parent store exists.
//
// `Relay` is the late-binding cell of the `AnyActionHandler` idiom: hand a
// handler forward where the receiver exists at hand-off, and create a relay
// where construction order prevents that. `bindSink` builds an
// `AnyActionHandler`, so the weak-owner capture is written once in this module.

/// A settable event funnel: `send` forwards to whatever `sink` holds at that moment.
/// Events before wiring are dropped by design — composition-root construction is
/// synchronous, so nothing real can fire in the gap.
public final class Relay<Event> {
  public var sink: ((Event) -> Void)?

  public init() {}

  public func send(_ event: Event) {
    sink?(event)
  }

  /// Sets `sink` to a closure that holds `owner` weakly and passes it back to
  /// `handler`. Events sent after `owner` deallocates are dropped, the same way
  /// events sent before wiring are.
  ///
  /// This is the sink form for the case where the capture IS an owner object —
  /// the store, shell, or worker whose mount the wiring belongs to. The weak hold
  /// keeps the mount free of an ARC cycle: the owner holds the relay it wired, and
  /// the relay does not hold the owner back. (The Kotlin twin's weak hold buys a
  /// different thing — there it keeps a torn-down shell collectible rather than
  /// reachable from a live relay.)
  ///
  /// ```swift
  /// routeRelay.bindSink(self) { root, route in root.store.send(.routed(route)) }
  /// ```
  ///
  /// Assign `sink` directly when the capture is not an owner object — a closure
  /// over values, or a strong capture the wiring deliberately wants.
  public func bindSink<Owner: AnyObject>(
    _ owner: Owner,
    _ handler: @escaping (Owner, Event) -> Void
  ) {
    let forwarder = AnyActionHandler<Event>(owner, closure: handler)
    sink = { event in forwarder.invoke(event) }
  }
}

File: kotlin/shells-compose/src/main/kotlin/dev/modaal/duet/shells/Relay.kt
// Copyright (c) 2026 Modaal.dev
// Licensed under the MIT License. See LICENSE file for details.

package dev.modaal.duet.shells

import java.lang.ref.WeakReference

// Breaks the child-environment ↔ parent-store construction cycle without
// late-init contortions. The composition root creates the relay, hands `send`
// to the child environment (or worker) it is building, and assigns `sink` once
// the parent store exists. Also the sanctioned ingress for app-level
// orchestration events emitted by workers — the same
// delegate-events-as-actions machinery children use.

/**
 * A settable event funnel: `send` forwards to whatever `sink` holds at that
 * moment. Events before wiring are dropped by design — composition-root
 * construction is synchronous, so nothing real can fire in the gap.
 * Swift-flavor mirror: DuetShells/Relay.swift.
 */
class Relay<Event> {
  var sink: ((Event) -> Unit)? = null

  fun send(event: Event) {
    sink?.invoke(event)
  }

  /**
   * Sets [sink] to a lambda that holds [owner] through a [WeakReference] and
   * passes it back to [handler]. Events sent after [owner] is collected are
   * dropped, the same way events sent before wiring are.
   *
   * This is the sink form for the case where the capture IS an owner object —
   * the store, shell, or worker whose mount the wiring belongs to. The weak
   * hold keeps a torn-down owner collectible: a relay that outlives its owner
   * holds no strong reference back. (The Swift twin's weak hold buys a
   * different thing — there it breaks an ARC cycle across the mount.)
   *
   * ```kotlin
   * routeRelay.bindSink(this) { root, route -> root.store.send(Routed(route)) }
   * ```
   *
   * Assign [sink] directly when the capture is not an owner object —
   * `routeRelay.sink = routeStore::send`, or a lambda over values.
   */
  fun <Owner : Any> bindSink(owner: Owner, handler: (Owner, Event) -> Unit) {
    val reference = WeakReference(owner)
    sink = { event -> reference.get()?.let { handler(it, event) } }
  }
}
